src/v/cloud_storage/inventory/inv_consumer.h

```diff
@@ -0,0 +1,122 @@
+      size_t max_hash_size_in_memory);
+
+    /// Consumes the report provided as input stream and populates in memory
+    /// path hashes. A flush to disk may be performed if the size of hashes
+    /// exceeds the prescribed limit.
+    ss::future<>
+    consume(ss::input_stream<char> s, is_gzip_compressed is_report_compressed);
+
+    static std::optional<model::ntp> ntp_from_path(std::string_view path);
+    static std::vector<ss::sstring> parse_row(ss::sstring row);
+
+    /// Ensures that streams opened by consumer are closed. Called during
+    /// exception handling if an exception is thrown during a stream
+    /// consumption.
+    ss::future<> stop();
+
+    ~inventory_consumer() = default;
+
+private:
+    ss::future<> process_paths(fragmented_vector<ss::sstring> paths);
+
+    // Checks if the path belongs to one of the NTPs whose leadership belongs to
+    // this node. If so, the path is hashed and added to the `_path_hashes` map.
+    void process_path(ss::sstring path);
+
+    // Writes the largest hash vectors to disk. The vectors are written in files
+    // named after their NTP. If write_all_entries is true, all hashes are
+    // written to disk. If it is not true, we write the vectors until total
+    // memory used falls below the max hash size limit.
+    ss::future<> flush(write_all_t write_all_entries = write_all_t::no);
+
+    // Writes hashes for a single NTP to a data file. The data file is
+    // located in path: namespace/topic/partition_id/{seq}. The parent
+    // directories are created if missing. A new data file is created for each
+    // flush operation.
+    ss::future<> flush_ntp_hashes(
+      model::ntp ntp, fragmented_vector<uint64_t> hashes, uint64_t file_name);
+    ss::future<>
+    write_hashes_to_file(ss::file& f, fragmented_vector<uint64_t> hashes);
+
+    ss::future<> write_hashes_to_file(
+      ss::output_stream<char>& stream, fragmented_vector<uint64_t> hashes);
+
+    // The root directory for hash files
+    std::filesystem::path _hash_store_path;
+
+    // NTPs for which leaders are shards on this node. The data produced by this
+    // class is expected to be used by the scrubbers which run on this node.
+    // Since scrubbing for an NTP is done by its partition leader, we need data
+    // for all possible scrubs done via this node, so we collect the NTPs led by
+    // this node up front and only process rows related to these.
+    absl::node_hash_set<model::ntp> _ntps;
+
+    struct flush_state {
+        uint64_t next_file_name{0};
+        fragmented_vector<uint64_t> hashes{};
+    };
+    // Mapping of hash vectors per NTP, populated as paths from inventory are
+    // processed.
+    absl::node_hash_map<model::ntp, flush_state> _ntp_flush_states;
+
+    // Max size of path hashes (8 bytes each) held in memory before a flush is
+    // performed to free up memory.
+    size_t _max_hash_size_in_memory;
+
+    // Total size of path hashes held in memory
+    size_t _total_size{};
```

Comment:
nit: can you add, is this in bytes? or number of entries?

Reply:
added comment in https://github.com/redpanda-data/redpanda/pull/21497 (this is in bytes)